selfdrive/common/spinner.c

```diff
@@ -166,11 +181,29 @@ int spin(int argc, char** argv) {
           bar_pos, progress_height-2, 12);
       nvgFillPaint(vg, paint);
       nvgFill(vg);
-    } else {
+    }
+
+    if (!draw_progress) {
       // message
       nvgTextAlign(vg, NVG_ALIGN_CENTER | NVG_ALIGN_TOP);
       nvgFontSize(vg, 96.0f);
       nvgText(vg, fb_w/2, (fb_h*2/3)+24, spintext, NULL);
+    } else if (has_extra) {
+      nvgTextAlign(vg, NVG_ALIGN_CENTER | NVG_ALIGN_TOP);
+      if (err_msg) {
+        int break_row_width = 1300;
+        int y_offset = strlen(spinerr) > 160 ? 76 : 96;
+        // need smaller font for longer error msg
+        int fontsize = strlen(spinerr) > 50 ? 68.0f : 72.0f;
```

Comment:
Could need some tuning. If there's an error with a large single chunk (no spaces) it will go off screen, I currently have no way to know how many rows nanovg is splitting text into

Reply:
You can split the lines in python. If you put in `\n` nanovg will correctly render them as linebreaks.

Reply:
I tried that but I think in spinner.c, this line breaks it: `spintext[strcspn(spintext, "\n")] = 0;` It renders `\n` weirdly, flashing the two lines when a new update is received. Is that `\n` only meant for the last one?

Reply:
That is to make sure the last `\n` is cut off.

Reply:
I wrote some code to cut off the last char (` = '0';`) if it's a `\n` in spinner.c however it still does not play nicely with newlines added in from Python, flashes on the screen. With `nvgTextBox` at least. I'll try `nvgText` and manually create new lines

Reply:
Got it working by replacing `\n` with the ASCII unit separator and then converting it back inside spinner to newlines! Now getting the line splitting logic ready. I also noticed I forgot to add `line = 'ERR,' + line` inside the formatter function so I'll add that when I get this done